Spectests: Use a flat structure for faster test ingestion

prestonvanloon opened the pull request:
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This PR is based on #15312. Please review and merge #15312 before reviewing this PR. 
This PR will remain a draft until #15312 merges.

**Which issues(s) does this PR fix?**

Each test target that ingests the spectest data corpus has to build a runfile tree. This is due to the use of `glob` with wildcards rather than hardcoding the files in BUILD. The runfile tree building operation is very expensive and continues to get more expensive as more directories and files are added.

**Other notes for review**

Here are the test results...

Before=>40 minutes to run spectests from a new machine.
After=7 minutes to run spectests from a new machine.

Reviewers:
- There is no logic change in any of the spectest files. These should be simple file move/rename.
- Please review `testing/spectest/README.md`

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

prestonvanloon committed:
Fix panic when checking types. String is not compatible with DeepEqual.

prestonvanloon committed:
Docs: add commentary on the filename convention

prestonvanloon committed:
Add a section about nightly tests to the spectest readme. Ref https://github.com/OffchainLabs/prysm/pull/15312

prestonvanloon committed:
Set shard_count to optimal value... one!

prestonvanloon committed:
Changelog fragment

kasey committed:
use latest unclog release

kasey commented on testing/spectest/README.md:
````diff
@@ -14,3 +14,24 @@ automatically when `//...` is selected. One can run minimal tests manually, thou
 ```bash
 bazel query 'tests(attr("tags", "minimal, spectest", //...))' | xargs bazel test --define ssz=minimal
````
Incidentally I think this is out of date and now you have to do this?
```
bazel query 'tests(attr("tags", "minimal, spectest", //...))' | xargs bazel test --//proto:network=minimal
```
Because when you try to run 
```
bazel query 'tests(attr("tags", "minimal, spectest", //...))' | xargs bazel test --define ssz=minimal
```
it gives you an error and basically tells you to do the thing I pasted above.

prestonvanloon replied:
Yes, you are correct. I'll fix

prestonvanloon replied:
You actually don't even need any of this. Minimal does build automatically with `//...` these days. I fixed that a while ago in #9122

kasey commented on testing/spectest/README.md:
````diff
@@ -14,3 +14,24 @@ automatically when `//...` is selected. One can run minimal tests manually, thou
+```
+
+An example test is the phase0 epoch processing test for effective balance updates. This test has a spectest path of `{mainnet, minimal}/phase0/epoch_processing/effective_balance_updates/pyspec_tests`.
+There are tests for mainnet and minimal config, so for each config we will add a file by the name of `phase0_epoch_processing_effective_balance_updates_test.go` since the fork is `phase0`, the package is `epoch_processing`, and the test is `effective_balance_updates`.
````
Since the spectest files use underscores in the file names, the underscore is an ambiguous replacement for slashes. I think it could be more clear if you used dashes instead, like 
```
phase0-epoch_processing-effective_balance_updates_test.go
```
Alternatively double underscore like:
```
phase0__epoch_processing__effective_balance_updates_test.go
```

kasey replied:
I assume these are mechanical renames. Could you provide the command you used? Then as the reviewer I could run that command on the same parent commit and confirm there is no change.

For extra credit, I think you can cleanly rebase `1f55ed77cd31479acb33d5650421310adfce8a3f` (where all the file moves happen) to be the tip commit. That would make it extra simple for me to run your command on the parent of the rename commit, and then a diff against this branch should show no changes?

kasey replied:
Would it be possible to somehow parameterize these tests so that we can filter which tests we run at once? Maybe with build tags? Presently I can for instance only run the forkchoice or ssz_static tests, but I think this change breaks that natural "filtering".

prestonvanloon replied:
The command I ran:

```
find . -type f -name '*_test.go' -exec bash -c '
  for file; do
    dir=$(dirname "$file")
    base=$(basename "$file" _test.go)
    new_name="./${dir#./}_${base}_test.go"
    mv
 "$file" "$new_name"
  done
' bash {}
```

I think i ran it twice

jtraglia commented:
Just want to mention that this PR is necessary for me to run these tests on a GitHub CI runner. The version in `develop` runs out of disk space and takes forever. This PR is much, much faster & doesn't use as much disk.

prestonvanloon committed:
Update spectest build instructions after #9122

prestonvanloon commented:
@jtraglia - Thanks for the context. 

I'll address @kasey feedback today and we will have this merged by EOW.

prestonvanloon committed:
Flatten spectests directory: Move all spectests to a single directory per network.

Commands ran:
```
cd testing/spectest/
```

Then for each network, I ran the following command twice.

```
find . -type f -name '*_test.go' -exec bash -c '
  for file; do
    dir=$(dirname "$file")
    base=$(basename "$file" _test.go)
    new_name="./${dir#./}__${base}_test.go"
    git mv "$file" "$new_name"
  done
' bash {} +
```

Then updated the packages with a command like

```
sed -i 's/package [a-zA-Z0-9_]\+/package mainnet/g' *.go
```

Updated commit from 5edadd7b to address @Kasey's feedback.

prestonvanloon commented:
@kasey I have made the changes in the filenames and documentation to reflect a double underscore separation

prestonvanloon commented:
I've run this spectest report and it is reporting many tests run successfully. The missing tests are mostly expected. 

[report.txt](https://github.com/user-attachments/files/20494497/report.txt)
